compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -1261,6 +1266,14 @@ impl VariantDef {
         Ident::new(self.name, tcx.def_ident_span(self.def_id).unwrap())
     }
 
+    pub fn fields(&self) -> &IndexVec<FieldIdx, FieldDef> {
+        if let Ok(fields) = &self.fields { fields } else { EMPTY_FIELDS }
+    }
+
+    pub fn fields_checked(&self) -> Result<&IndexVec<FieldIdx, FieldDef>, ErrorGuaranteed> {
+        self.fields.as_ref().map_err(|e| *e)
+    }
```

Comment:
I think if we do decide that we should go ahead with this strategy, we should still keep `fields` as public on `VariantDef` and expose some other way for `hir_typeck` to detect that the ADT/Variant has been tainted.

An idea:
0. Store `tainted: Option<ErrorGuaranteed>` in the ADT, revert the changes to the fields list
1. Expose `VariantDef::has_errors() -> Result<(), ErrorGuaranteed>`
2. Make the `TyCtxt::type_of` query return `TyKind::Error` for ADTs that have recovered.